### ISSUE TITLE
[java] New rule: TypeNameMismatch

### ISSUE THREAD

**Issue by zbynek** [pmd/pmd#7033]
**Proposed Rule Name:** TypeNameMismatch

**Proposed Category:** Code Style

**Description:** 
Top  level types should be defined in a file with a matching name. For public types this is enforced by the compiler, but for package-private types it's possible to have a mismatch. That can be confusing for developers who expect to be able to locate a source file based on class name.

**Code Sample:** This should include code, that should be flagged by the rule. If possible, the "correct" code
according to this new rule should also be demonstrated.

```java
// content of Foo.java

class Bar { // violation, should be defined in Bar.java

}
```

```java
// content of Baz.java
class Baz {
  class Foo { } // OK, nested classes are allowed
}

class Bar { // violation, second top-level class should be in its own file
}
```

**Possible Properties:**

None.

**Comment by UncleOwen** [comment]
What do you think about checking the directory structure as well? For example, making sure that a class in the `foo.bar.baz` package is located in a directory called `.../foo/bar/baz`?
